### PR TITLE
[go_router] implemented helpers for StatefulShellRoute

### PR DESCRIPTION
**Pull request by hangyujin**
issue: [#127371](https://github.com/flutter/flutter/issues/127371)

go_router_builder_pr  https://github.com/flutter/packages/pull/4238 

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by chunhtai** [comment]
Can you also link a go_router_builder pr?

**Comment by hangyujin** [comment]
> 

Yes, but it's still a draft because I need to merge the go_router pr first and update the dependency in go_router_builder

**Comment by chunhtai** [comment]
that is fine, it is easier to review after knowing the whole picture

File: packages/go_router/lib/src/route_data.dart [hunk 1]
```diff
@@ -205,6 +205,109 @@ abstract class ShellRouteData extends RouteData {
...
+    required T Function(GoRouterState) factory,
+    required List<StatefulShellBranch> branches,
+    String? restorationScopeId,
+  }) {
+    T factoryImpl(GoRouterState state) {
+      final Object? extra = state.extra;
+
+      // If the "extra" value is of type `T` then we know it's the source
+      // instance of `StatefulShellRouteData`, so it doesn't need to be recreated.
+      if (extra is T) {
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:245]: This would never be true i think?

File: packages/go_router/lib/src/route_data.dart [hunk 2]
```diff
@@ -205,6 +205,109 @@ abstract class ShellRouteData extends RouteData {
...
+/// Base class for supporting
+/// [StatefulShellRoute](https://pub.dev/documentation/go_router/latest/go_router/StatefulShellRoute-class.html)
+abstract class StatefulShellBranchData extends RouteData {
+  /// Default const constructor
+  const StatefulShellBranchData();
+
+  /// A helper function used by generated code.
+  ///
+  /// Should not be used directly.
+  static StatefulShellBranch $route<T extends StatefulShellBranchData>({
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:299]: Its weird we still call this $route

File: packages/go_router/lib/src/route_data.dart [hunk 3]
```diff
@@ -205,6 +205,109 @@ abstract class ShellRouteData extends RouteData {
...
+  /// to minimize the number of times it has to be deserialized.
+  static final Expando<StatefulShellRouteData> _stateObjectExpando =
+      Expando<StatefulShellRouteData>(
+    'GoRouteState to StatefulShellRouteData expando',
+  );
+}
+
+/// Base class for supporting
+/// [StatefulShellRoute](https://pub.dev/documentation/go_router/latest/go_router/StatefulShellRoute-class.html)
+abstract class StatefulShellBranchData extends RouteData {
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:292]: This probably should be its own type, I don't think we should reuse the RouteData.

Comment (chunhtai) [reply packages/go_router/lib/src/route_data.dart:292]: It can't reuse some of the logic from RouteData. The way to navigate to a branch is to use goBranch.

File: packages/go_router/lib/src/route_data.dart [hunk 4]
```diff
@@ -256,6 +359,36 @@ class TypedShellRoute<T extends ShellRouteData> extends TypedRoute<T> {
   final List<TypedRoute<RouteData>> routes;
 }
 
+/// A superclass for each typed shell route descendant
+@Target(<TargetKind>{TargetKind.library, TargetKind.classType})
+class TypedStatefulShellRoute<T extends StatefulShellRouteData>
+    extends TypedRoute<T> {
+  /// Default const constructor
+  const TypedStatefulShellRoute({
+    this.routes = const <TypedRoute<RouteData>>[],
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:368]: this should call branches?

File: packages/go_router/lib/src/route_data.dart (hunk 1 above)
Comment (chunhtai) [reply packages/go_router/lib/src/route_data.dart:245]: This applies to all other RouteData class

File: packages/go_router/lib/src/route_data.dart [hunk 5]
```diff
@@ -205,6 +205,109 @@ abstract class ShellRouteData extends RouteData {
...
+    StatefulNavigationShell navigationShell,
+  ) =>
+      throw UnimplementedError(
+        'One of `builder` or `pageBuilder` must be implemented.',
+      );
+
+  /// A helper function used by generated code.
+  ///
+  /// Should not be used directly.
+  static StatefulShellRoute $route<T extends StatefulShellRouteData>({
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:235]: I think you can refactor most of the duplicated logic from ShellRouteData to a ShellRouteBaseData

File: packages/go_router/lib/src/route_data.dart [hunk 6]
```diff
@@ -205,6 +205,109 @@ abstract class ShellRouteData extends RouteData {
...
+      GoRouterState state,
+      StatefulNavigationShell navigationShell,
+    ) =>
+        factoryImpl(state).pageBuilder(
+          context,
+          state,
+          navigationShell,
+        );
+
+    return StatefulShellRoute.indexedStack(
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:274]: since you are using indexedStack, maybe call this class IndexedStackShellRouteData

File: packages/go_router/lib/src/route_data.dart (hunk 4 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route_data.dart]: done

File: packages/go_router/lib/src/route_data.dart (hunk 3 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route_data.dart]: done

File: packages/go_router/lib/src/route_data.dart (hunk 5 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route_data.dart:235]: Actually most logics are slightly different, `ShellRouteData` use a Widget in pagebuilder and builder and `StatefulShellRouteData` use a StatefulNavigationShell

```
Widget builder(
      BuildContext context,
      GoRouterState state,
      Widget navigator,
    ) 
```
```
Widget builder(
      BuildContext context,
      GoRouterState state,
      StatefulNavigationShell navigationShell,  //<------here
    ) 
```

File: packages/go_router/lib/src/route_data.dart (hunk 6 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route_data.dart:275]: update to only fallback to indexedStack  when `navigatorContainerBuilder`  is null

File: packages/go_router/lib/src/route_data.dart (hunk 2 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route_data.dart]: renamed to $branch

File: packages/go_router/lib/src/route_data.dart (hunk 1 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route_data.dart]: deleted

File: packages/go_router/lib/src/route.dart [hunk 7]
```diff
@@ -655,8 +655,8 @@ class StatefulShellRoute extends ShellRouteBase {
     required this.navigatorContainerBuilder,
     this.restorationScopeId,
   })  : assert(branches.isNotEmpty),
-        assert((pageBuilder != null) ^ (builder != null),
```
Comment (chunhtai) [inline packages/go_router/lib/src/route.dart:658]: why this change?

File: packages/go_router/lib/src/route_data.dart [hunk 8]
```diff
@@ -159,7 +159,7 @@ abstract class ShellRouteData extends RouteData {
       final Object? extra = state.extra;
 
       // If the "extra" value is of type `T` then we know it's the source
-      // instance of `GoRouteData`, so it doesn't need to be recreated.
+      // instance of `ShellRouteData`, so it doesn't need to be recreated.
       if (extra is T) {
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:163]: This can be removed?

File: packages/go_router/lib/src/route_data.dart (hunk 5 above)
Comment (chunhtai) [reply packages/go_router/lib/src/route_data.dart:235]: sounds good

File: packages/go_router/test/route_data_test.dart [hunk 9]
```diff
@@ -181,6 +249,46 @@ void main() {
...
+      'It should build the page from the overridden build method',
+      (WidgetTester tester) async {
+        final GoRouter goRouter = GoRouter(
+          initialLocation: '/child',
+          routes: <RouteBase>[
+            _statefulShellRouteDataBuilder,
+          ],
+        );
+        await tester.pumpWidget(MaterialApp.router(
+          routeInformationProvider: goRouter.routeInformationProvider,
```
Comment (chunhtai) [inline packages/go_router/test/route_data_test.dart:263]: Use `MaterialApp.router(routerConfig: goRouter)` instead. Can you also convert any other place that you found doesn't use router config?

File: packages/go_router/test/route_data_test.dart [hunk 10]
```diff
@@ -86,6 +86,74 @@ final ShellRoute _shellRouteDataPageBuilder = ShellRouteData.$route(
...
+}
+
+class _StatefulShellRouteDataBuilder extends StatefulShellRouteData {
+  const _StatefulShellRouteDataBuilder();
+
+  @override
+  Widget builder(
+    BuildContext context,
+    GoRouterState state,
+    Widget navigator,
```
Comment (chunhtai) [inline packages/go_router/test/route_data_test.dart:100]: this should be StatefulNavigationShell?

**Review by chunhtai — COMMENTED** [review]
I think most of the code LGTM, can you also update the go_router_builder pr so that i can review as a whole?

File: packages/go_router/lib/src/route.dart (hunk 7 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route.dart:658]: StatefulShellRouteData.$route will provide both pageBuilder and builder, those builder might return `null` but the builder itself is not null,

**Comment by hangyujin** [comment]
updated the go_router_builder pr

File: packages/go_router/lib/src/route_data.dart [hunk 11]
```diff
@@ -205,6 +197,109 @@ abstract class ShellRouteData extends RouteData {
...
+        builder: builder,
+        pageBuilder: pageBuilder,
+        navigatorContainerBuilder: navigatorContainerBuilder,
+      );
+    }
+    return StatefulShellRoute.indexedStack(
+      branches: branches,
+      builder: builder,
+      pageBuilder: pageBuilder,
+      restorationScopeId: restorationScopeId,
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:271]: this should also be in line 260, also how would one specify this in TypeStatefulShellRoute?

Comment (chunhtai) [reply packages/go_router/lib/src/route_data.dart:271]: we should also consider adding this to TypedShellRoute

File: packages/go_router/lib/src/route_data.dart [hunk 12]
```diff
@@ -205,6 +197,109 @@ abstract class ShellRouteData extends RouteData {
...
+
+  /// A helper function used by generated code.
+  ///
+  /// Should not be used directly.
+  static StatefulShellBranch $branch<T extends StatefulShellBranchData>({
+    GlobalKey<NavigatorState>? navigatorKey,
+    List<RouteBase> routes = const <RouteBase>[],
+  }) {
+    return StatefulShellBranch(
+      routes: routes,
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:297]: how does one specify initialLocation and restorationScopeId? we may also want a way to specify observers

File: packages/go_router/lib/src/route_data.dart (hunk 11 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route_data.dart:276]: done

File: packages/go_router/lib/src/route_data.dart (hunk 12 above)
Comment (hangyujin) [reply packages/go_router/lib/src/route_data.dart:305]: done

**Review by chunhtai — APPROVED** [review]
LGTM
